### ISSUE TITLE
Discovery / Search Feature 

### ISSUE THREAD

**Issue by ebresie** [jenv/jenv#362]
Presently, jenv requires to manually "add" each jdk path on the system.

It would be nice to have a jenv discover feature which would search the system (maybe can specify a specific path to reduce the search), to discover available JDKs present.

So if there is a collection of different java versions in a given folder, find and add each of the JDK folders. Maybe with each find could have an "autogenerated" id/label and/or path shown and ask if would like to add or not for use in an interactive mode.

May have to take care also to not add the same version multiple times so check if given one exists already from the list of know/previously added jdks.

---
This is based on a similar feature request on Jenv for windows.  See https://github.com/FelixSelter/JEnv-for-Windows/issues/32

**Comment by marquiswang** [comment]
Agreed that this would be very nice. Would just need a list of paths of known locations for JDKs for various OSes.

For Macs:

```
/Library/Java/JavaVirtualMachines/*/Contents/Home
/usr/local/Cellar/openjdk*/
/opt/homebrew/Cellar/openjdk*/
```

Linux (Redhat/Debian distros anyways):
```
/usr/lib/jvm/*
```

Do you know of any other paths to look in?

**Comment by andrewflbarnes** [comment]
Had a quick look at autoscan in windows jenv - would definitely be a nice feature to have, especially for new users. This there would probably be a couple of variants.
- `jenv autoscan` - search well known locations
- `jenv autoscan <folder>` - search the given folder
- `jenv autoscan --all` - search the whole filesystem

Might also be useful to have options like
- `-y` - auto add any versions found if not already linked

Would need to determine/effect the below
- order of precedence of locations
- I think
- do we follow symlinks as part of the search to other folders?
- probably shouldn't scan nfs? maybe there should be an option like `--allow-nfs`/`--allow-all`
- probably shouldn't scan external? maybe there should be an option like `--allow-external`/`--allow-all`
- what locations do we always avoid (`/proc`, `/dev`, etc.)?

As a first cut can probably just implement the well known locations search and explicit folder search and prevent the user from supplying `/`.

**Comment by divStar** [comment]
Just some ideas (this ticket is already old 😄 ):

- Start simple, meaning: either make the auto-scanned paths **configurable** via some config file or hard-code them
- It would be helpful to allow **wildcards** for more performant scanning (so instead of having to use `/opt/homebrew/Cellar`, I'd gladly use `/opt/homebrew/Cellar/openjdk*/` which should include `/opt/homebrew/Cellar/openjdk` as well as `/opt/homebrew/Cellar/openjdk@17` for example)
- Perhaps make following symlinks and scanning external volumes and NFS storage also configurable?

Doing a complete system scan is pretty pointless in my opinion, because usually people do not just have their JDKs/JREs randomly scattered on their system. So by default this function could be disabled altogether - with perhaps a hint to some docs (from the `jenv -h` usage output), that help one set up the auto-scanning.

This probably applies to MacOS and Linux. On Windows the JDKs/JREs are probably either in `C:\Program Files\` or a dedicated folder. It'd still work with the suggestion above though, but we need to be careful about special directory permissions (in all OSes).

My issue currently is, that if I do `brew upgrade` and it upgrades all my packages, the old `openjdk` folders usually are not valid anymore, because they've gotten upgrades (minor or revision). However: `jenv` attaches a JDK to a `<major>.<minor>.<revision>` and `<major>.<minor>` version (as well as just `<major>`), which is all good, but if I just add the new version, I only update the most precise one, not all the others (MacOS):
```shell
igvo@~  jenv add /opt/homebrew/Cellar/openjdk/23.0.2/libexec/openjdk.jdk/Contents/Home/
openjdk64-23.0.2 added
23.0.2 added
 23.0 already present, skip installation
 23 already present, skip installation
```
Note the `already present, skip installation` lines; this means, that if I use `jenv local 23` somewhere, it'll be pointing to a non-existent JDK.

So I have to do this:
```shell
igvo@~  jenv remove 23.0.2
JDK 23.0.2 removed
igvo@~  jenv remove 23.0  
JDK 23.0 removed
igvo@~  jenv remove 23
```
before
```shell
igvo@~  jenv add /opt/homebrew/Cellar/openjdk/23.0.2/libexec/openjdk.jdk/Contents/Home/   
 openjdk64-23.0.2 already present, skip installation
23.0.2 added
23.0 added
23 added
```
and this is makes it more difficult than necessary in my opinion.
